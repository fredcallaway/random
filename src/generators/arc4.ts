import type { Seed } from '../types'
import { RNG } from '../rng'
import { mixKey } from '../utils'

//
// ARC4
//
// An ARC4 implementation. The constructor takes a key in the form of an array
// of at most (width) integers that should be 0 <= x < (width).
//
// The g(count) method returns a pseudorandom integer that concatenates the
// next (count) outputs from ARC4. Its return value is a number x that is in
// the range 0 <= x < (width ^ count).

// The following constants are related to IEEE 754 limits.
// const width = 256 // each RC4 output is 0 <= x < 256
// const chunks = 6 // at least six RC4 outputs for each double
const _arc4_startdenom = 281_474_976_710_656 // 256 ** 6 == width ** chunks
const _arc4_significance = 4_503_599_627_370_496 // 2 ** 52 significant digits in a double
const _arc4_overflow = 9_007_199_254_740_992 // 2 ** 53 == significance * 2

export class ARC4RNG extends RNG {
  protected readonly _seed: Seed

  i: number
  j: number
  S: number[]

  constructor(seed: Seed = crypto.randomUUID()) {
    super()

    this._seed = seed
    const key = mixKey(seed, [])

    const S: number[] = []
    const keylen = key.length
    this.i = 0
    this.j = 0
    this.S = S

    // Set up S using the standard key scheduling algorithm.
    let i = 0
    while (i <= 0xff) {
      S[i] = i++
    }

    for (let i = 0, j = 0; i <= 0xff; i++) {
      const t = S[i]!
      j = 0xff & (j + key[i % keylen]! + t)
      S[i] = S[j]!
      S[j] = t
    }

    // For more robust unpredictability, the function call below discards an
    // initial batch of values. This is called RC4-drop.
    this.g(256)
  }

  override get name() {
    return 'arc4'
  }

  override next() {
    // This function returns a random double in [0, 1) that contains
    // randomness in every bit of the mantissa of the IEEE 754 value.

    let n = this.g(6) // Start with a numerator n < 2 ^ 48
    let d = _arc4_startdenom // and denominator d = 2 ^ 48.
    let x = 0 // and no 'extra last byte'.

    while (n < _arc4_significance) {
      // Fill up all significant digits (2 ** 52)
      n = (n + x) * 256 // by shifting numerator and
      d *= 256 // denominator and generating a
      x = this.g(1) // new least-significant-byte.
    }

    while (n >= _arc4_overflow) {
      // To avoid rounding past overflow, before adding
      n /= 2 // last byte, shift everything
      d /= 2 // right using integer math until
      x >>>= 1 // we have exactly the desired bits.
    }

    return (n + x) / d // Form the number within [0, 1).
  }

  g(count: number) {
    const { S } = this
    let { i, j } = this
    let r = 0

    while (count--) {
      i = 0xff & (i + 1)
      const t = S[i]!
      S[j] = t
      j = 0xff & (j + t)
      S[i] = S[j]!
      r = r * 256 + S[0xff & (S[i]! + t)]!
    }

    this.i = i
    this.j = j

    return r
  }

  override clone() {
    return new ARC4RNG(this._seed)
  }
}
