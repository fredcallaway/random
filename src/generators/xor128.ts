import type { Seed } from '../types'
import { RNG } from '../rng'

export class XOR128RNG extends RNG {
  protected readonly _seed: Seed

  x: number
  y: number
  z: number
  w: number

  constructor(seed: Seed = crypto.randomUUID()) {
    super()

    this._seed = seed
    this.x = 0
    this.y = 0
    this.z = 0
    this.w = 0

    let strSeed: string = ''

    if (typeof seed === 'number') {
      this.x = seed
    } else {
      strSeed += `${seed}`
    }

    // Mix in string seed, then discard an initial batch of 64 values.
    for (let i = 0; i < strSeed.length + 64; ++i) {
      this.x ^= strSeed.charCodeAt(i) | 0
      this.next()
    }
  }

  override get name() {
    return 'xor128'
  }

  override next() {
    const t = this.x ^ (this.x << 11)
    this.x = this.y
    this.y = this.z
    this.z = this.w
    this.w = this.w ^ ((this.w >>> 19) ^ t ^ (t >>> 8))
    return (this.w >>> 0) / 0x1_00_00_00_00
  }

  override clone() {
    return new XOR128RNG(this._seed)
  }
}
