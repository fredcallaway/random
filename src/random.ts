import type { RNG } from './rng'
import type { SeedOrRNG } from './types'
import { bates } from './distributions/bates'
import { bernoulli } from './distributions/bernoulli'
import { binomial } from './distributions/binomial'
import { exponential } from './distributions/exponential'
import { geometric } from './distributions/geometric'
import { irwinHall } from './distributions/irwin-hall'
import { logNormal } from './distributions/log-normal'
import { normal } from './distributions/normal'
import { pareto } from './distributions/pareto'
import { poisson } from './distributions/poisson'
import { uniform } from './distributions/uniform'
import { uniformBoolean } from './distributions/uniform-boolean'
import { uniformInt } from './distributions/uniform-int'
import { MathRandomRNG } from './generators/math-random'
import { createRNG } from './utils'

/**
 * Distribution function
 */
type IDistFn<R> = (random: Random, ...argv: any) => R

/**
 * Distribution
 */
type IDist<R> = () => R

/**
 * Keyed cache entry
 */
interface ICacheEntry<T> {
  key: string
  distribution: () => T
}

/**
 * Seedable random number generator supporting many common distributions.
 *
 * @name Random
 * @class
 *
 * @param {RNG|function|string|number} [rng=Math.random] - Underlying random number generator or a seed for the default PRNG. Defaults to `Math.random`.
 */
export class Random {
  protected _rng!: RNG
  protected readonly _cache: {
    [k: string]: ICacheEntry<any>
  } = {}

  constructor(seedOrRNG: SeedOrRNG = new MathRandomRNG()) {
    this._rng = createRNG(seedOrRNG)
  }

  /**
   * @member {RNG} rng - Underlying pseudo-random number generator.
   */
  get rng() {
    return this._rng
  }

  /**
   * Creates a new `Random` instance, optionally specifying parameters to
   * set a new seed.
   *
   * @return {Random}
   */
  clone(seedOrRNG: SeedOrRNG = this.rng.clone()): Random {
    return new Random(seedOrRNG)
  }

  /**
   * Sets the underlying pseudorandom number generator.
   *
   * @example
   * ```ts
   * import random from 'random'
   *
   * random.use('example-seed')
   * // or
   * random.use(Math.random)
   * ```
   */
  use(seedOrRNG: SeedOrRNG) {
    this._rng = createRNG(seedOrRNG)
  }

  // --------------------------------------------------------------------------
  // Uniform utility functions
  // --------------------------------------------------------------------------

  /**
   * Convenience wrapper around `this.rng.next()`
   *
   * Returns a floating point number in [0, 1).
   *
   * @return {number}
   */
  next = (): number => {
    return this._rng.next()
  }

  /**
   * Samples a uniform random floating point number, optionally specifying
   * lower and upper bounds.
   *
   * Convenience wrapper around `random.uniform()`
   *
   * @param {number} [min=0] - Lower bound (float, inclusive)
   * @param {number} [max=1] - Upper bound (float, exclusive)
   * @return {number}
   */
  float = (min?: number, max?: number): number => {
    return this.uniform(min, max)()
  }

  /**
   * Samples a uniform random integer, optionally specifying lower and upper
   * bounds.
   *
   * Convenience wrapper around `random.uniformInt()`
   *
   * @param {number} [min=0] - Lower bound (integer, inclusive)
   * @param {number} [max=1] - Upper bound (integer, inclusive)
   * @return {number}
   */
  int = (min?: number, max?: number) => {
    return this.uniformInt(min, max)()
  }

  /**
   * Samples a uniform random integer, optionally specifying lower and upper
   * bounds.
   *
   * Convenience wrapper around `random.uniformInt()`
   *
   * @alias `random.int`
   *
   * @param {number} [min=0] - Lower bound (integer, inclusive)
   * @param {number} [max=1] - Upper bound (integer, inclusive)
   * @return {number}
   */
  integer = (min?: number, max?: number) => {
    return this.uniformInt(min, max)()
  }

  /**
   * Samples a uniform random boolean value.
   *
   * Convenience wrapper around `random.uniformBoolean()`
   *
   * @alias `random.boolean`
   *
   * @return {boolean}
   */
  bool = () => {
    return this.uniformBoolean()()
  }

  /**
   * Samples a uniform random boolean value.
   *
   * Convenience wrapper around `random.uniformBoolean()`
   *
   * @return {boolean}
   */
  boolean = () => {
    return this.uniformBoolean()()
  }

  /**
   * Returns an item chosen uniformly at random from the given array.
   *
   * Convenience wrapper around `random.uniformInt()`
   *
   * @param {Array<T>} [array] - Input array
   * @return {T | undefined}
   */
  choice<T>(array: Array<T>): T | undefined {
    if (!Array.isArray(array)) {
      throw new TypeError(
        `Random.choice expected input to be an array, got ${typeof array}`
      )
    }

    const length = array.length

    if (length > 0) {
      const index = this.uniformInt(0, length - 1)()
      return array[index]
    } else {
      return undefined
    }
  }
  /**
   * Returns a shuffled copy of the given array.
   *
   * @param {Array<T>} [array] - Input array
   * @return {Array<T>}
   */
  shuffle<T>(array: Array<T>): Array<T> {
    if (!Array.isArray(array)) {
      throw new TypeError(
        `Random.shuffle expected input to be an array, got ${typeof array}`
      )
    }
    const copy = [...array]
    ShuffleInPlace(this.rng, copy)
    return copy
  }
  /**
   * Generates a thunk which returns shuffled copies of the given array.
   *
   * @param {Array<T>} [array] - Input array
   * @return {function}
   */
  shuffler<T>(array: Array<T>): () => Array<T> {
    if (!Array.isArray(array)) {
      throw new TypeError(
        `Random.shuffler expected input to be an array, got ${typeof array}`
      )
    }
    const gen = this.rng
    const copy = [...array]
    return () => {
      ShuffleInPlace(gen, copy)
      return [...copy]
    }
  }
  // --------------------------------------------------------------------------
  // Uniform distributions
  // --------------------------------------------------------------------------

  /**
   * Generates a [Continuous uniform distribution](https://en.wikipedia.org/wiki/Uniform_distribution_(continuous)).
   *
   * @param {number} [min=0] - Lower bound (float, inclusive)
   * @param {number} [max=1] - Upper bound (float, exclusive)
   */
  uniform = (min?: number, max?: number) => {
    return this._memoize<number>('uniform', uniform, min, max)
  }

  /**
   * Generates a [Discrete uniform distribution](https://en.wikipedia.org/wiki/Discrete_uniform_distribution).
   *
   * @param {number} [min=0] - Lower bound (integer, inclusive)
   * @param {number} [max=1] - Upper bound (integer, inclusive)
   * @return {function}
   */
  uniformInt = (min?: number, max?: number) => {
    return this._memoize<number>('uniformInt', uniformInt, min, max)
  }

  /**
   * Generates a [Discrete uniform distribution](https://en.wikipedia.org/wiki/Discrete_uniform_distribution),
   * with two possible outcomes, `true` or `false.
   *
   * This method is analogous to flipping a coin.
   *
   * @return {function}
   */
  uniformBoolean = () => {
    return this._memoize<boolean>('uniformBoolean', uniformBoolean)
  }

  // --------------------------------------------------------------------------
  // Normal distributions
  // --------------------------------------------------------------------------

  /**
   * Generates a [Normal distribution](https://en.wikipedia.org/wiki/Normal_distribution).
   *
   * @param {number} [mu=0] - Mean
   * @param {number} [sigma=1] - Standard deviation
   * @return {function}
   */
  normal = (mu?: number, sigma?: number) => {
    return normal(this, mu, sigma)
  }

  /**
   * Generates a [Log-normal distribution](https://en.wikipedia.org/wiki/Log-normal_distribution).
   *
   * @param {number} [mu=0] - Mean of underlying normal distribution
   * @param {number} [sigma=1] - Standard deviation of underlying normal distribution
   * @return {function}
   */
  logNormal = (mu?: number, sigma?: number) => {
    return logNormal(this, mu, sigma)
  }

  // --------------------------------------------------------------------------
  // Bernoulli distributions
  // --------------------------------------------------------------------------

  /**
   * Generates a [Bernoulli distribution](https://en.wikipedia.org/wiki/Bernoulli_distribution).
   *
   * @param {number} [p=0.5] - Success probability of each trial.
   * @return {function}
   */
  bernoulli = (p?: number) => {
    return bernoulli(this, p)
  }

  /**
   * Generates a [Binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution).
   *
   * @param {number} [n=1] - Number of trials.
   * @param {number} [p=0.5] - Success probability of each trial.
   * @return {function}
   */
  binomial = (n?: number, p?: number) => {
    return binomial(this, n, p)
  }

  /**
   * Generates a [Geometric distribution](https://en.wikipedia.org/wiki/Geometric_distribution).
   *
   * @param {number} [p=0.5] - Success probability of each trial.
   * @return {function}
   */
  geometric = (p?: number) => {
    return geometric(this, p)
  }

  // --------------------------------------------------------------------------
  // Poisson distributions
  // --------------------------------------------------------------------------

  /**
   * Generates a [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution).
   *
   * @param {number} [lambda=1] - Mean (lambda > 0)
   * @return {function}
   */
  poisson = (lambda?: number) => {
    return poisson(this, lambda)
  }

  /**
   * Generates an [Exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution).
   *
   * @param {number} [lambda=1] - Inverse mean (lambda > 0)
   * @return {function}
   */
  exponential = (lambda?: number) => {
    return exponential(this, lambda)
  }

  // --------------------------------------------------------------------------
  // Misc distributions
  // --------------------------------------------------------------------------

  /**
   * Generates an [Irwin Hall distribution](https://en.wikipedia.org/wiki/Irwin%E2%80%93Hall_distribution).
   *
   * @param {number} [n=1] - Number of uniform samples to sum (n >= 0)
   * @return {function}
   */
  irwinHall = (n?: number) => {
    return irwinHall(this, n)
  }

  /**
   * Generates a [Bates distribution](https://en.wikipedia.org/wiki/Bates_distribution).
   *
   * @param {number} [n=1] - Number of uniform samples to average (n >= 1)
   * @return {function}
   */
  bates = (n?: number) => {
    return bates(this, n)
  }

  /**
   * Generates a [Pareto distribution](https://en.wikipedia.org/wiki/Pareto_distribution).
   *
   * @param {number} [alpha=1] - Alpha
   * @return {function}
   */
  pareto = (alpha?: number) => {
    return pareto(this, alpha)
  }

  // --------------------------------------------------------------------------
  // Internal
  // --------------------------------------------------------------------------

  /**
   * Memoizes distributions to ensure they're only created when necessary.
   *
   * Returns a thunk which that returns independent, identically distributed
   * samples from the specified distribution.
   *
   * @internal
   *
   * @param {string} label - Name of distribution
   * @param {function} getter - Function which generates a new distribution
   * @param {...*} args - Distribution-specific arguments
   *
   * @return {function}
   */
  protected _memoize<T>(
    label: string,
    getter: IDistFn<any>,
    ...args: any[]
  ): IDist<T> {
    const key = `${args.join(';')}`
    let value = this._cache[label]

    if (value === undefined || value.key !== key) {
      value = {
        key,
        distribution: getter(this, ...args)
      }
      this._cache[label] = value
    }

    return value.distribution
  }
}

// Helper function
function ShuffleInPlace<T>(gen: RNG, array: Array<T>) {
  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(gen.next() * (i + 1))
    const tmp = array[i]
    array[i] = array[j] as T
    array[j] = tmp as T
  }
}

export default new Random()
