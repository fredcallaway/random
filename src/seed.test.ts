import seedrandom from 'seedrandom'
import { assert, expect, test } from 'vitest'

import random, { Random } from '../src/random'
import { XOR128RNG } from './generators'

test('random.clone with seedrandom rng is consistent', () => {
  const r = random.clone(seedrandom('ZjExZDczNWQxY2NlZjUzYmRiZWU0ZGIz'))
  const d = r.uniform()
  const o: number[] = []
  for (let i = 0; i < 10; ++i) {
    const v = d()
    o.push(v)
  }

  expect(o).toMatchSnapshot()
})

test('random.clone with string seed is consistent', () => {
  const r = random.clone('MDUzNzM5ZGQwYWZlNWIzNjI5MDYwNWNm')
  const d = r.normal()
  const o: number[] = []
  for (let i = 0; i < 10; ++i) {
    const v = d()
    o.push(v)
  }

  expect(o).toMatchSnapshot()
})

test('Random constructor', () => {
  const rng = new Random()
  expect(rng.rng.name).toEqual('Math.random')

  const rng2 = new Random(seedrandom('my-seed-string'))
  expect(rng2.rng.name).toEqual('prng')

  const rng3 = new Random(Math.random)
  expect(rng3.rng.name).toEqual('random')

  const rng4 = new Random('example-seed-string')
  expect(rng4.rng.name).toEqual('arc4')
})

test('random seed consistency', () => {
  let value = 0

  for (let i = 0; i < 10; ++i) {
    const rng = new Random(seedrandom('my-seed'))
    if (i <= 0) {
      value = rng.float()
    } else {
      assert.equal(value, rng.float())
    }
  }

  for (let i = 0; i < 10; ++i) {
    const rng = new Random('my-seed')
    if (i <= 0) {
      value = rng.float()
    } else {
      assert.equal(value, rng.float())
    }
  }

  for (let i = 0; i < 10; ++i) {
    const rng = new Random(42)
    if (i <= 0) {
      value = rng.float()
    } else {
      assert.equal(value, rng.float())
    }
  }

  for (let i = 0; i < 10; ++i) {
    const rng = new Random('25')
    if (i <= 0) {
      value = rng.float()
    } else {
      assert.equal(value, rng.float())
    }
  }
})

test('random.use() different seeds produce different values', () => {
  const n = 1100
  const length = 10

  random.use('25')
  const values0 = Array.from({ length }).map(() => random.int(n))

  random.use('abcdef')
  const values1 = Array.from({ length }).map(() => random.int(n))
  assert.notDeepEqual(values0, values1)

  random.use('abcdefxyz')
  const values2 = Array.from({ length }).map(() => random.int(n))
  assert.notDeepEqual(values0, values2)
  assert.notDeepEqual(values1, values2)

  random.use('25xyz')
  const values3 = Array.from({ length }).map(() => random.int(n))
  assert.notDeepEqual(values0, values3)
  assert.notDeepEqual(values1, values3)
  assert.notDeepEqual(values2, values3)

  random.use('25')
  const values4 = Array.from({ length }).map(() => random.int(n))
  assert.deepEqual(values0, values4)

  random.use('abcdef')
  const values5 = Array.from({ length }).map(() => random.int(n))
  assert.deepEqual(values1, values5)
})

test('random.use() XOR128RNG different seeds produce different values', () => {
  const r = new Random()
  const n = 1100
  const length = 10

  r.use(new XOR128RNG('25'))
  const values0 = Array.from({ length }).map(() => r.int(n))

  r.use(new XOR128RNG('abcdef'))
  const values1 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values1)

  r.use(new XOR128RNG('abcdefxyz'))
  const values2 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values2)
  assert.notDeepEqual(values1, values2)

  r.use(new XOR128RNG('25xyz'))
  const values3 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values3)
  assert.notDeepEqual(values1, values3)
  assert.notDeepEqual(values2, values3)

  r.use(new XOR128RNG('25'))
  const values4 = Array.from({ length }).map(() => r.int(n))
  assert.deepEqual(values0, values4)

  r.use(new XOR128RNG('abcdef'))
  const values5 = Array.from({ length }).map(() => r.int(n))
  assert.deepEqual(values1, values5)
})

test('new Random() uses a different seed each time', () => {
  const n = 1100
  const length = 10

  let r = new Random()
  const values0 = Array.from({ length }).map(() => r.int(n))

  r = new Random()
  const values1 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values1)

  r = new Random()
  const values2 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values2)
  assert.notDeepEqual(values1, values2)

  r = new Random()
  const values3 = Array.from({ length }).map(() => r.int(n))
  assert.notDeepEqual(values0, values3)
  assert.notDeepEqual(values1, values3)
  assert.notDeepEqual(values2, values3)
})
