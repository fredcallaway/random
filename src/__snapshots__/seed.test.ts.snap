// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`random.clone with seedrandom rng is consistent 1`] = `
[
  0.9362808130793461,
  0.4864517867266793,
  0.5065797249862649,
  0.3300815411147545,
  0.022876541849241913,
  0.0018718398576695902,
  0.8458705577357042,
  0.9501863429937075,
  0.5562872420975038,
  0.8032030618791965,
]
`;

exports[`random.clone with string seed is consistent 1`] = `
[
  0.2944232742184869,
  0.35115469131789284,
  0.14051693549513977,
  -0.5556060076011705,
  -0.4790220609212565,
  3.0110380960265593,
  2.1567558775511335,
  -2.052822767200372,
  -0.8166047767287957,
  -0.40142278586882546,
]
`;
